contracts/src/BoundlessMarket.sol

```diff
@@ -224,78 +225,113 @@ contract BoundlessMarket is
+                    selectors: assessorFill.selectors,
+                    root: claimDigest,
+                    prover: assessorFill.prover
+                })
+            )
+        );
+        // Verification that the provided seal matches the required selector.
+        if (assessorFill.selectors.length == 1 && assessorFill.selectors[0].value != bytes4(fill.seal[0:4])) {
+            revert SelectorMismatch(assessorFill.selectors[0].value, bytes4(fill.seal[0:4]));
+        }
         // Verification of the assessor seal does not need to comply with FULFILL_MAX_GAS_FOR_VERIFY.
-        VERIFIER.verify(assessorSeal, ASSESSOR_ID, assessorJournalDigest);
+        VERIFIER.verify(assessorFill.seal, ASSESSOR_ID, assessorJournalDigest);
     }
 
     /// @inheritdoc IBoundlessMarket
-    function verifyBatchDelivery(Fulfillment[] calldata fills, bytes calldata assessorSeal, address prover)
-        public
-        view
-    {
+    function verifyBatchDelivery(Fulfillment[] calldata fills, FulfillmentAssessor calldata assessorFill) public view {
         // TODO(#242): Figure out how much the memory here is costing. If it's significant, we can do some tricks to reduce memory pressure.
-        bytes32[] memory claimDigests = new bytes32[](fills.length);
-        bytes32[] memory requestDigests = new bytes32[](fills.length);
-        for (uint256 i = 0; i < fills.length; i++) {
-            requestDigests[i] = fills[i].requestDigest;
-            claimDigests[i] = ReceiptClaimLib.ok(fills[i].imageId, sha256(fills[i].journal)).digest();
-            VERIFIER.verifyIntegrity{gas: FULFILL_MAX_GAS_FOR_VERIFY}(Receipt(fills[i].seal, claimDigests[i]));
+        uint256 fillsLength = fills.length;
+        // We can't handle more than 65535 fills in a single batch.
+        // This is a limitation of the current Selectors implementation, and can be increased in the future.
+        if (fillsLength > type(uint16).max) {
+            revert BatchSizeExceedsLimit(fillsLength, type(uint16).max);
+        }
+        bytes32[] memory claimDigests = new bytes32[](fillsLength);
+        bytes32[] memory requestDigests = new bytes32[](fillsLength);
+
+        uint256 selectorsLength = assessorFill.selectors.length;
+        uint16 selectorIdx = 0;
+
+        for (uint256 i = 0; i < fillsLength; i++) {
+            Fulfillment calldata fill = fills[i];
+
+            requestDigests[i] = fill.requestDigest;
+            claimDigests[i] = ReceiptClaimLib.ok(fill.imageId, sha256(fill.journal)).digest();
+
+            // If the current index is flagged for selector verification, process it.
+            if (selectorIdx < selectorsLength && assessorFill.selectors[selectorIdx].index == i) {
+                if (assessorFill.selectors[selectorIdx].value != bytes4(fill.seal[0:4])) {
+                    revert SelectorMismatch(assessorFill.selectors[selectorIdx].value, bytes4(fill.seal[0:4]));
+                }
+                selectorIdx++;
+            }
```

Comment:
If `assessorFill.selectors[selectorIdx]` is not sorted properly, some of the selector constraints will never be applied.